src/zkevm_specs/evm/execution/sar.py

```diff
@@ -0,0 +1,172 @@
+from ...util import FQ, N_BYTES_U64, RLC
+from ..instruction import Instruction, Transition
+from ..opcode import Opcode
+from ..typing import Sequence
+
+MAX_U64 = 0xFFFFFFFFFFFFFFFF
+
+
+def sar(instruction: Instruction):
+    opcode = instruction.opcode_lookup(True)
+
+    a = instruction.stack_pop()
+    shift = instruction.stack_pop()
+    b = instruction.stack_push()
+
+    (
+        a64s,
+        b64s,
+        a64s_lo,
+        a64s_hi,
+        shf_div64,
+        shf_mod64,
+        p_lo,
+        p_hi,
+        p_top,
+        is_neg,
+    ) = __gen_witness(instruction, opcode, a, shift)
+    __check_witness(
+        instruction,
+        a,
+        shift,
+        b,
+        a64s,
+        b64s,
+        a64s_lo,
+        a64s_hi,
+        shf_div64,
+        shf_mod64,
+        p_lo,
+        p_hi,
+        p_top,
+        is_neg,
+    )
+
+    instruction.step_state_transition_in_same_context(
+        opcode,
+        rw_counter=Transition.delta(2),
+        program_counter=Transition.delta(1),
+        stack_pointer=Transition.delta(1),
+    )
+
+
+def __check_witness(
+    instruction: Instruction,
+    a: RLC,
+    shift: RLC,
+    b: RLC,
+    a64s: Sequence[FQ],
+    b64s: Sequence[FQ],
+    a64s_lo: Sequence[FQ],
+    a64s_hi: Sequence[FQ],
+    shf_div64,
+    shf_mod64,
+    p_lo,
+    p_hi,
+    p_top,
+    is_neg,
+):
+    shf_lt256 = instruction.is_zero(instruction.sum(shift.le_bytes[1:]))
```

Comment:
If we take the sum of the bytes from `idx=1`, we should then do a decomposition check for `shift` so that `shf_lt256` is correctly constrained.

Reply:
It's not very clear here, can you give me more context?
There is already `shift` constraint [here](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/586275ff2c04c8beea0730f9034f3de052183b1b/src/zkevm_specs/evm/execution/sar.py#L121-L125), is it enough

Reply:
Fixed to use `CompareWordGadget` for `shf_lt256` as:
```
shf_lt256, _ = instruction.compare_word(shift, RLC(256))
```

Reply:
Replaced `shf_lt256` with `shf_hi_is_zero` (`IsZeroGadget`).
In `gen_witness`:
```
shf_lo, shf_hi = word_to_lo_hi(shift)
shf_div64 = shf_lo.n // 64
shf_mod64 = shf_lo.n % 64
...
```
And in `check_witness`:
```
shf_lo, shf_hi = instruction.word_to_lo_hi(shift)
shf_hi_is_zero = instruction.is_zero(shf_hi)
```